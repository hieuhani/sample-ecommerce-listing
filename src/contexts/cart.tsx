import React, { useReducer, useContext, useMemo } from 'react'
import { Product } from '../repositories/products'

export interface CartProviderProps {
  children: React.ReactNode;
}

export interface Cart {
  [key: string]: {
    quantity: number;
    price: number;
  };
}

export interface State {
  cart: Cart;
}

export type Action =
  | { type: 'ADD'; payload: Product }
  | { type: 'DELETE'; payload: Product };

type Dispatch = (action: Action) => void

const initialState: State = {
  cart: {}
}

export type CartState = State & {
  totalItems: number;
  totalValue: number;
}

const CartStateContext = React.createContext<CartState>(initialState as CartState)
const CartDispatchContext = React.createContext<Dispatch | undefined>(undefined)

export function cartReducer(state: State, action: Action): State {
  switch (action.type) {
    case 'ADD': {
      const product = action.payload
      return {
        ...state,
        cart: Object.assign({}, state.cart, {
          [product.sku]: {
            price: product.price.sellPrice,
            quantity: state.cart[product.sku] ? state.cart[product.sku].quantity + 1 : 1,
          }
        })
      }
    }
    case 'DELETE': {
      const product = action.payload
      if (!state.cart[product.sku]) {
        return state
      }
      if (state.cart[product.sku].quantity > 1) {
        return {
          ...state,
          cart: Object.assign({}, state.cart, {
            [product.sku]: {
              price: product.price.sellPrice,
              quantity: state.cart[product.sku].quantity - 1,
            }
          })
        }
      }
      const { [product.sku]: value, ...cart } = state.cart
      return {
        ...state,
        cart,
      }
    }
  }
}

export function CartProvider({ children }: CartProviderProps): React.ReactNode {
  const [state, dispatch] = useReducer(cartReducer, initialState)
  const cartState: CartState = useMemo(() => {
    let totalItems = 0
    let totalValue = 0
    Object.keys(state.cart).forEach(productSku => {
      totalItems += state.cart[productSku].quantity
      totalValue += state.cart[productSku].quantity * state.cart[productSku].price
    })
    return {
      cart: state.cart,
      totalItems,
      totalValue,
    }
  }, [state])
  return (
    <CartStateContext.Provider value={cartState}>
      <CartDispatchContext.Provider value={dispatch}>
        {children}
      </CartDispatchContext.Provider>
    </CartStateContext.Provider>
  )
}

export function useCartState(): CartState {
  const context = useContext(CartStateContext)
  if (!context) {
    throw new Error('useCartState must be inside a CartProvider with a state value')
  }
  return context
}

export function useCartDispatch(): Dispatch {
  const context = useContext(CartDispatchContext)
  if (!context) {
    throw new Error('useCartDispatch must be inside a CartProvider with a state dispatch')
  }
  return context
}
