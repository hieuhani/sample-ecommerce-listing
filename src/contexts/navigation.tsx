import React, { useReducer, useContext, useMemo } from 'react'
import { ScreenPayload } from '../types'

export interface NavigationProviderProps {
  children: React.ReactNode;
}

export interface State {
  stack: Array<ScreenPayload>;
}

export type Action =
  | { type: 'PUSH'; payload: ScreenPayload }
  | { type: 'POP' };

type Dispatch = (action: Action) => void

const initialState: State = {
  stack: [{
    name: 'PRODUCT_LISTING',
  }]
}

export type ScreenPayloadWithOrder = ScreenPayload & {
  order: number;
}

export type NavigationState = State & {
  currentScreen: ScreenPayload | undefined;
  stackWithOrder: Array<ScreenPayloadWithOrder>;
}

const NavigationStateContext = React.createContext<NavigationState>(initialState as NavigationState)
const NavigationDispatchContext = React.createContext<Dispatch | undefined>(undefined)

export function navigationReducer(state: State, action: Action): State {
  switch (action.type) {
    case 'PUSH':
      return {
       ...state,
       stack: [...state.stack, action.payload],
      }
    case 'POP':
      return {
        ...state,
        stack: state.stack.length > 1 ? state.stack.filter((_, index) => index !== state.stack.length - 1) : initialState.stack
      }
    default:
      throw new Error()
  }
}

export function NavigationProvider({ children }: NavigationProviderProps): React.ReactNode {
  const [state, dispatch] = useReducer(navigationReducer, initialState)
  const navigationState: NavigationState = useMemo(() => {
    return {
      stack: state.stack,
      currentScreen: state.stack.length > 0 ? state.stack.slice(-1)[0] : undefined,
      stackWithOrder: state.stack.map((item, index) => Object.assign({}, item, {
        order: index,
      }))
    }
  }, [state])
  return (
    <NavigationStateContext.Provider value={navigationState}>
      <NavigationDispatchContext.Provider value={dispatch}>
        {children}
      </NavigationDispatchContext.Provider>
    </NavigationStateContext.Provider>
  )
}

export function useNavigationState(): NavigationState {
  const context = useContext(NavigationStateContext)
  if (!context) {
    throw new Error('useNavigationState must be inside a NavigationProvider with a state value')
  }
  return context
}

export function useNavigationDispatch(): Dispatch {
  const context = useContext(NavigationDispatchContext)
  if (!context) {
    throw new Error('useNavigationDispatch must be inside a NavigationProvider with a state dispatch')
  }
  return context
}
