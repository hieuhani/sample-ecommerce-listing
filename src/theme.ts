import { DefaultTheme } from 'styled-components'

export const theme: DefaultTheme = {
  colors: {
    reddishOrange: '#f5471e',
    coolGrey: '#8f9598',
    silver: '#c6cace',
    tomato: '#ea341f',
    darkGrey: '#262829',
    paleGrey: '#eef1f3',
    white: '#FFF',
    black30: 'rgba(0, 0, 0, .3)',
    pumpkinOrange: '#f58400',
    deepSkyBlue: '#1e75f6',
  }
}
