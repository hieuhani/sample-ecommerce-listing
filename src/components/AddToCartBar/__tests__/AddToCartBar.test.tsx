import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { AddToCartBar } from '..'
import { theme } from '../../../theme'
import { CartProvider } from '../../../contexts/cart'

const props = {
  cartState: {
    cart: {},
    totalItems: 0,
    totalValue: 0,
  },
  product: {
    displayName: 'Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)',
    sku: '1807564',
    images: [{
      url: 'https://phongvu.vn/media/catalog_v2/media/48/18/1570775921.343882_Acer_AS_XC-885_1.jpg',
      priority: 1,
    }],
    status: {
      publish: true,
      sale: 'ngung_kinh_doanh' as const,
    },
    promotionPrices: [{
      channel: 'pv_online',
      terminal: 'phongvu',
      finalPrice: 11890000,
      bestPrice: 11890000,
    }],
    price: {
      supplierSalePrice: 11890000,
      sellPrice: 11890000,
    },
    saleCategories: [],
    attributeGroups: [],
    attributes: [],
  },
}
describe('AddToCartBar', () => {
  test('AddToCartBar renders correctly', () => {
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <CartProvider>
          <AddToCartBar {...props} />
        </CartProvider>
      </ThemeProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
