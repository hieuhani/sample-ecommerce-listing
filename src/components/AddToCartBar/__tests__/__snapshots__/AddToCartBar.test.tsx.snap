// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToCartBar AddToCartBar renders correctly 1`] = `
.c1 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgb(232,32,32);
  background: linear-gradient(90deg,rgba(232,32,32,1) 0%,#f5471e 100%);
  border-radius: 8pt;
  color: #FFF;
  padding: 8pt;
  max-width: 250pt;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 span {
  margin-left: 50pt;
  font-size: 15pt;
}

.c0 {
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  padding: 8pt;
  background-color: #FFF;
  z-index: 99999;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div
  className="c0"
>
  <button
    className="c1"
    onClick={[Function]}
  >
    <img
      src="/public/images/AddToCart.png"
      srcSet="/public/images/AddToCart@2x.png 2x, /public/images/AddToCart@3x.png 3x"
    />
    <span>
      Add To Cart
    </span>
  </button>
</div>
`;
